react-custom-hooks/client/src/components/useTodos.js

```diff
@@ -0,0 +1,115 @@
+/* eslint-disable no-unused-vars -- Remove me */
+import { useEffect, useState } from 'react';
+
+const url = (path) => `${process.env.REACT_APP_BASE_URL}${path}`;
+
+/**
+ * Manages the Todos by reading from and writing to the backend API.
+ *
+ * @returns {Object} with the following properties:
+ *   `isLoading`: a boolean indicating if the data has not yet loaded;
+ *   `todos`: a list of To Do items;
+ *   `addTodo`: a function (newTodo) that will add a new To Do;
+ *   `toggleCompleted`: a function (todoId) that will toggle the completion of an existing To Do.
+ */
+export function useTodos() {
+  const [todos, setTodos] = useState();
+  //My added below
+  useEffect(() => {
+    /* TODO: If `todos` hasn't been defined yet, read the items from the API
+     * and set them into the `todos` state.
+     */
+    async function loadThis() {
+      if (!todos) {
+        let readThis = await readTodos()
+        setTodos(readThis)
+      }
+    }
+    loadThis()
+
+  }, [todos]);
+
+  function addTodo(newTodo) {
+    /* TODO: Call the API function that creates a Todo item.
+     * When the promise returned by that function resolves, update the `todos` state.
+     * Note that it is critical that you pass a _new_ array. Do not modify the `todos` array.
+     */
+
+    async function addThis() {
+      const todo = await createTodo(newTodo)
+      console.log(todo)
+      setTodos((prev) => prev.concat(todo))
+    }
+    addThis()
+  }
+
+  function toggleCompleted(todoId) {
+    const findTodo = todos.find((todo) => todo.todoId === todoId)
+    console.log(findTodo)
+    if (findTodo) {
+      findTodo.isCompleted = !findTodo.isCompleted
+    }
+    async function updateThis() {
+      const editedTodo = await updateTodo(findTodo)
+      console.log('This is for toggleCompleted', editedTodo)
+      setTodos((prev) => prev.concat(editedTodo))
```

Comment:
This isn't the right way to do this. It will add the edited Todo to the end of the list. Since it forces a redraw, you get the correct list eventually in the useEffect, but you should be making a copy of the array and _replacing_ the item you just edited, rather than concatenating it.